motionplan/inverseKinematics.go

```diff
@@ -2,23 +2,260 @@ package motionplan
 
 import (
 	"context"
+	"encoding/json"
+	"math"
+	"runtime"
+	"sort"
+
+	"github.com/edaniels/golog"
+	"go.viam.com/utils"
 
 	"go.viam.com/rdk/referenceframe"
-	spatial "go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/spatialmath"
+)
+
+// default values for inverse kinematics.
+const (
+	// If an IK solution scores below this much, return it immediately.
+	defaultMinIkScore = 0.
+
+	// Number of IK solutions that should be generated before stopping.
+	defaultSolutionsToSeed = 50
 )
 
-// InverseKinematics defines an interface which, provided with a goal position and seed inputs, will output all found
-// solutions to the provided channel until cancelled or otherwise completes.
-type InverseKinematics interface {
-	// Solve receives a context, the goal arm position, and current joint angles.
-	Solve(context.Context, chan<- []referenceframe.Input, spatial.Pose, []referenceframe.Input, Metric, int) error
+var defaultNumThreads = runtime.NumCPU() / 2
+
+type ikOptions struct {
+	constraintHandler
+	extra map[string]interface{}
+
+	// Metric by which to measure nearness to the goal
+	metric Metric
+
+	// Solutions that score below this amount are considered "good enough" and returned immediately
+	MinScore float64 `json:"min_ik_score"`
+
+	// For the below values, if left uninitialized, default values will be used. To disable, set < 0
+	// Max number of ik solutions to consider
+	MaxSolutions int `json:"max_ik_solutions"`
+
+	// Number of cpu cores to use
+	NumThreads int `json:"num_threads"`
 }
 
-func limitsToArrays(limits []referenceframe.Limit) ([]float64, []float64) {
-	var min, max []float64
-	for _, limit := range limits {
-		min = append(min, limit.Min)
-		max = append(max, limit.Max)
+func newBasicIKOptions() *ikOptions {
```

Comment:
A potential usability improvement would be passing in the `ikConfig` map as an argument and handling that before you initially return `opts. 

There's precedent for doing this AND adding the JSON marshal/unmarshal code, such as what is done with cbirrt options. Compare L97-104 in this file versus L55-62 of cBiRRT.go in the newCbirrtOptions method). Not critical, but might save repeating a few lines in newIKSolver.

Reply:
I think this is mostly taken care of in the latest update